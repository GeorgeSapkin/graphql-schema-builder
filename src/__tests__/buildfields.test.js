'use strict';

const graphql = require('graphql');

const {
  buildFields
} = require('..')(graphql);

const {
  assetResolvers,
  assetResolversWithArgs,
  assetResolversWithArgsAsFunc,
  AssetWithMeasurements,
  Customer,
  CustomerNested,
  customerResolvers,
  nop,
  schemaStore
} = require('./__fixtures__');

const {
  types: {
    ObjectId
  }
} = require('../types');

describe('buildFields', () => {
  describe('returns a field schema', () => {
    const getExistingType = schemaStore.get.bind(schemaStore);

    it('from object fields', () => {
      const allFields = {
        ...Customer.fields,
        ...Customer.dynamicFields({ ObjectId })
      };

      expect(
        buildFields(allFields, {
          getExistingType,

          resolvers: customerResolvers
        })
      ).toMatchSnapshot();
    });

    it('from function fields', () => {
      expect(
        buildFields(Customer.dynamicFields, {
          getExistingType,

          resolvers: customerResolvers
        })
      ).toMatchSnapshot();
    });

    it('from a field with an object resolver without args', () => {
      expect(
        buildFields(AssetWithMeasurements.fields, {
          getExistingType,

          resolvers: assetResolvers
        })
      ).toMatchSnapshot();
    });

    it('from a field with an object resolver with object args', () => {
      expect(
        buildFields(AssetWithMeasurements.fields, {
          getExistingType,

          resolvers: assetResolversWithArgs
        })
      ).toMatchSnapshot();
    });

    it('from a field with an object resolver with function args', () => {
      expect(
        buildFields(AssetWithMeasurements.fields, {
          getExistingType,

          resolvers: assetResolversWithArgsAsFunc
        })
      ).toMatchSnapshot();
    });

    it('from nested fields', () => {
      const fields = buildFields(CustomerNested.fields);
      expect(fields.metadata.type).toMatchSnapshot();
    });
  });

  describe('throws', () => {
    it('without fields', () => expect(buildFields).toThrow());

    it('without buildSubType', () => expect(() => buildFields({}, {
      buildSubType: null
    })).toThrow());
    it('without buildSubType', () => expect(() => buildFields({}, {
      buildSubType:    nop,
      getExistingType: null
    })).toThrow());
  });
});
