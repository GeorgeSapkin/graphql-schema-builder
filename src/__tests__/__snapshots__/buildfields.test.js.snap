// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildFields returns a field schema from a field with an object resolver with function args 1`] = `
Object {
  "measurements": Object {
    "args": Object {
      "time": Object {
        "type": "DateTime",
      },
    },
    "resolve": [Function],
    "type": "[Measurement!]!",
  },
}
`;

exports[`buildFields returns a field schema from a field with an object resolver with object args 1`] = `
Object {
  "measurements": Object {
    "args": Object {
      "resolution": Object {
        "type": "String!",
      },
    },
    "resolve": [Function],
    "type": "[Measurement!]!",
  },
}
`;

exports[`buildFields returns a field schema from a field with an object resolver without args 1`] = `
Object {
  "measurements": Object {
    "resolve": [Function],
    "type": "[Measurement!]!",
  },
}
`;

exports[`buildFields returns a field schema from function fields 1`] = `
Object {
  "assets": Object {
    "resolve": [Function],
    "type": "[Asset!]!",
  },
}
`;

exports[`buildFields returns a field schema from nested fields 1`] = `"metadata"`;

exports[`buildFields returns a field schema from object fields 1`] = `
Object {
  "assets": Object {
    "resolve": [Function],
    "type": "[Asset!]!",
  },
  "name": Object {
    "description": "The name of the customer.",
    "type": "String!",
  },
  "value": Object {
    "type": "Float!",
  },
}
`;
